packages/indexer/indexer/db.py

```diff
@@ -0,0 +1,168 @@
+                    (NEW.block->'block'->'header'->>'proposer_address')::TEXT
+                );
+                
+                        
+                FOR tx_responses IN SELECT * FROM jsonb_array_elements(NEW.txs->'tx_responses')
+                LOOP
+                    INSERT INTO txs (txhash, chain_id, height, tx_response, tx, tx_response_tx_type, code, data, info, logs, events, raw_log, gas_used, gas_wanted, codespace, timestamp)
+                    VALUES (
+                        tx_responses->>'txhash',
+                        NEW.chain_id,
+                        NEW.height,
+                        tx_responses,
+                        tx_responses->'tx',
+                        tx_responses->'tx'->>'@type',
+                        tx_responses->>'code',
+                        tx_responses->>'data',
+                        tx_responses->>'info',
+                        tx_responses->'logs',
+                        tx_responses->'events',
+                        tx_responses->>'raw_log',
+                        (tx_responses->>'gas_used')::BIGINT,
+                        (tx_responses->>'gas_wanted')::BIGINT,
+                        tx_responses->>'codespace',
+                        (tx_responses->'timestamp')::TEXT::TIMESTAMP
+                    );
+                END LOOP;
+                
+                RETURN NEW;
+            END
+            $$ LANGUAGE plpgsql;
+            CREATE OR REPLACE TRIGGER raw_insert
+            BEFORE INSERT
+            ON raw
+            FOR EACH ROW EXECUTE PROCEDURE parse_raw();
+            """
+        )
+
+
+async def upsert_block(pool: asyncpg.pool, block: dict, txs: dict):
+    async with pool.acquire() as conn:
+        chain_id, height = block["block"]["header"]["chain_id"], int(
+            block["block"]["header"]["height"]
+        )
+        await conn.execute(
+            """
+            INSERT INTO raw (chain_id, height, block, txs)
+            VALUES ($1, $2, $3, $4)
+            ON CONFLICT DO NOTHING
+            """,
+            chain_id,
+            height,
+            json.dumps(block),
+            json.dumps(txs),
+        )
+        # for some reason this doesn't support params
+        await conn.execute(
+            f"""
+            NOTIFY raw, '{chain_id} {height}';
+            """
+        )
+
+
+async def get_missing_blocks(
+    pool: asyncpg.pool, session: aiohttp.ClientSession, chain: CosmosChain
+):
+
+    while True:
+        block_data = await get_block(session, chain.apis[0])
+        current_block = int(block_data["block"]["header"]["height"])
+        async with pool.acquire() as conn:
+            rows = await conn.fetch(
+                """
+                SELECT height FROM blocks WHERE chain_id = $1
+                """,
+                chain.chain_id,
+            )
+            heights = [row[0] for row in rows]
+            for height in range(chain.min_block_height, current_block):
```

Comment:
My concern is what if there is missing data in-between for some reason?

For example, say we have blocks `[1,2,3,4,5]`
then later end up in a state where we have `[1, 2, 3, 4, 5, 10, 11, 12, 13]` but the current blockheight is `20`. We want to make sure we don't query any blocks that we already have.

Reply:
it accounts for this:

min height in your example is 1 so you iterate from 1-20 removing those blocks in that list (look at sudocode below)

```
for every possible height in blockchain (from min height to current height):
    if this height hasn't queried
        upsert
```

Reply:
I still think we need to take some set difference or something. What if there are tens of millions of blocks? Seems silly to iterate over the entire list.

Reply:
Its a generator so it yields blocks as they are needed, this is more efficient than doing all of the computation at the beginning